### ISSUE TITLE
Training time explosion (caffe 0.16)

### ISSUE THREAD

**Issue by RSly** [NVIDIA/caffe#349]
Hi,

I am monitoring the training time of caffe 0.16 for a simple network on 4gpus and batch_size 16.
here you can see the training and testing time for the first two epochs.

=> for the first epoch, all is normal: the training takes around 2 seconds per iteration, and the test time is reasonable

=> after the first epoch, the training time explodes by 70 times!!

this looks like a very strange and harmful behavior!
can anyone explain this?

![image](https://user-images.githubusercontent.com/3428292/26916605-49d08ab4-4c2b-11e7-817c-146964d9b904.png)

**Comment by RSly** [comment]
hi @drnikolaev, these are maybe mini hangs remaining from https://github.com/NVIDIA/caffe/issues/315?

you can reproduce it by a simple python layer from  https://github.com/NVIDIA/DIGITS/tree/master/examples/medical-imaging#dice-metric

* gpu usages are at 0, for the majority of the training time in those 150-second iterations.

**Comment by RSly** [comment]
Hi @drnikolaev , I have prepared a minimal example so the problem can be reproduced easily :)

here is a solver, a pre_trained.caffemodel to initialize the net, and a minimal prototxt (intentionally there is no real loss layer, so mainly no initial weights will change in time)
https://drive.google.com/open?id=0Bzpjahml31eoWmJhdEZBVGhIRk0
layer "dicenv" is from  https://github.com/NVIDIA/DIGITS/tree/master/examples/medical-imaging#dice-metric

=> once using 4 gpus, the caffe_output log is:
I0608 18:48:46.896667  3826 solver.cpp:300] Iteration 70 (0.847392 iter/s, 11.8009s/10 iter), loss = 0
I0608 18:48:58.599087  3826 solver.cpp:300] Iteration 80 (0.854533 iter/s, 11.7023s/10 iter), loss = 0
I0608 18:49:05.011214  3826 solver.cpp:441] Iteration 84, Testing net (#0)
I0608 18:49:25.406738  3826 solver.cpp:259] [MultiGPU] Tests completed in 20.3957s
I0608 18:49:43.140538  3826 solver.cpp:300] Iteration 90 (0.224508 iter/s, 44.5418s/10 iter), loss = 0
I0608 18:55:19.979636  3826 solver.cpp:300] Iteration 100 (0.0296875 iter/s, 336.842s/10 iter), loss = 0
I0608 19:09:38.849555  3826 solver.cpp:300] Iteration 110 (0.0116431 iter/s, 858.877s/10 iter), loss = 0
I0608 19:09:38.850747  3826 solver.cpp:322]     Train net output #0: dicenv = 1
I0608 19:09:38.850764  3826 sgd_solver.cpp:108] Iteration 110, lr = 0.0001

=> at iteration 90, the processing time starts exploding, at iteration 110 it takes ~78 times of the usual training time (858 seconds instead of 11)

FYI:
* I can reproduce the problem on the dummy data from create_images in https://github.com/NVIDIA/caffe/issues/315#issuecomment-299246892
* Also it happens even if we iterate on only 1 training image!!
* tried with and without conv_algos_override: "1,1,1" => doesnt help
* tried with and without Engine: CAFFE => doesnt help

**Comment by drnikolaev** [comment]
Yeah, this is syncing issue, looking to it...
Thanks for the sample!

**Comment by drnikolaev** [comment]
@RSly I can't reproduce this, sorry. My timings are pretty much even. Perhaps I use some other dataset. Could you upload yours somewhere?

**Comment by RSly** [comment]
Hi @drnikolaev ,
Do you run your training for large number of images to be processed? 
I have noticed that the problem does not occur in the first epoch. 

Here is my timing curve in seconds/iteration, for two epochs on MS coco dataset with 80k images, 4 gpus, and bacthsize 16/GPU

the problem happens once around 60k images have been processed

![image](https://user-images.githubusercontent.com/3428292/27294469-f4a546ea-5519-11e7-8db4-64355a1f52e7.png)

**Comment by drnikolaev** [comment]
@RSly could you verify on your side whether your PythonLayer raises any exceptions?

**Comment by RSly** [comment]
hi @drnikolaev, no I dont see any exceptions. 

=> the gpu usage remains at 0% for 95% of those 200/400 seconds you see in the figure, and it goes to working at 80-100% of gpu usage only for 5-10 seconds; these 5-10 are actually where the iteration is processed

Q: To initialize the weights of a fcn alexnet, did you actually use my weights from https://drive.google.com/open?id=0Bzpjahml31eoWmJhdEZBVGhIRk0   ?
With these wights you get the problem already after some iterations



[update] I decided to dig deeper to benchmark the timing for both test and train, here you can see that it is far from regular and it can explode both for training iterations and testing by 30-100 times

1. Training time per **iteration** for 5k iterations

![image](https://user-images.githubusercontent.com/3428292/27333937-e7d098d0-55c7-11e7-91a3-42e32a2f5250.png)

2. Test time per **Epoch** (seconds) for 4 epochs

![image](https://user-images.githubusercontent.com/3428292/27334127-9334d8bc-55c8-11e7-9a06-1e261ee85195.png)

and here, test and train timings on the same figure: 
![image](https://user-images.githubusercontent.com/3428292/27335023-71d3f312-55cb-11e7-967b-e5d7f836434e.png)




**in conclusion:**  basically during those periods that this problem happens, it affects both testing and training times! now the only question is that what makes this happen? :)

**Comment by RSly** [comment]
hi @drnikolaev , a quick feedback to let you know that the time explosion is not present in 0.16.2, see in the next figure and compare it with figure 1 from the previous post!! magical :)

It can be very interested to find out what caused and solved this issue for the future developement.
do you know which part of the code could be the cause here?

thanks 

![image](https://user-images.githubusercontent.com/3428292/27579052-790557c4-5b26-11e7-8514-74501bd50165.png)

**Comment by drnikolaev** [comment]
Hi @RSly , thanks for confirming! We overhauled PythonLayer implementation and added more robust exception handling.
